src/sandbox/mod.rs

```diff
@@ -0,0 +1,176 @@
+//! Basic Pod Sandbox types
+
+pub mod pinned;
+
+use anyhow::Result;
+use derive_builder::Builder;
+use getset::Getters;
+use std::fmt;
+
+#[derive(Builder, Getters)]
+#[builder(pattern = "owned", setter(into))]
+pub struct Sandbox<T>
+where
+    T: Default,
+{
+    #[get = "pub"]
+    /// The unique identifier.
+    id: String,
+
+    #[get = "pub"]
+    /// Full name of the sandbox.
+    name: String,
+
+    #[get = "pub"]
+    /// Namespace where the sandbox lives in.
+    namespace: String,
+
+    #[get = "pub"]
+    /// Sandbox creation attempt
+    attempt: u32,
+
+    /// Trait implementation for creating the sandbox.
+    #[builder(default = "T::default()")]
+    implementation: T,
+}
+
+pub trait Pod {
+    /// Run a previously created sandbox.
+    fn run(&mut self) -> Result<()> {
+        Ok(())
+    }
+
+    /// Stop a previously started sandbox.
+    fn stop(&mut self) -> Result<()> {
+        Ok(())
+    }
+
+    /// Remove a stopped sandbox.
+    fn remove(&mut self) -> Result<()> {
+        Ok(())
+    }
+}
+
+impl<T> Sandbox<T>
+where
+    T: Default + Pod,
+{
+    /// Wrapper for the implementations `run` method
+    pub fn run(&mut self) -> Result<()> {
+        self.implementation.run()
+    }
+
+    #[allow(dead_code)]
+    /// Wrapper for the implementations `stop` method
+    pub fn stop(&mut self) -> Result<()> {
+        self.implementation.stop()
+    }
+
+    #[allow(dead_code)]
+    /// Wrapper for the implementations `remove` method
+    pub fn remove(&mut self) -> Result<()> {
+        self.implementation.remove()
+    }
+}
+
+impl<T> fmt::Debug for Sandbox<T>
+where
+    T: Default,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("Sandbox")
+            .field("id", &self.id())
+            .field("name", &self.name())
+            .field("namespace", &self.namespace())
+            .field("attempt", &self.attempt())
+            .finish()
+    }
+}
+
+impl<T> fmt::Display for Sandbox<T>
+where
+    T: Default,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{} ({})", self.name(), self.id())
+    }
+}
+
+#[cfg(test)]
```

Comment:
I don't know much about idiomatic rust, but do we maybe want to separate the test impl from the main module file?

Reply:
This is only possible if we put every module in a separate directory containing a mod.rs and test.rs. Possible, but then the structure of the project gets flooded by all those files having the same name. 😀